File: project_euler_unified/problem_109/solution.py
"""
Problem 109: Darts

Count distinct ways to check out (reach exactly 0) with score < 100.
Last dart must be a double. Up to 3 darts total.
Non-final darts are unordered (combinations, not permutations).
"""

def solve():
    # Build all 62 distinct darts
    # Each dart has a score and an index for canonical ordering
    all_darts = []  # (score, label) - label for identification

    # Singles: S1..S20, S25
    for i in range(1, 21):
        all_darts.append((i, f'S{i}'))
    all_darts.append((25, 'S25'))

    # Doubles: D1..D20, D25
    doubles = []
    for i in range(1, 21):
        all_darts.append((2 * i, f'D{i}'))
        doubles.append(2 * i)
    all_darts.append((50, 'D25'))
    doubles.append(50)

    # Trebles: T1..T20
    for i in range(1, 21):
        all_darts.append((3 * i, f'T{i}'))

    scores = [d[0] for d in all_darts]
    n = len(scores)  # 62

    count = 0

    # 1-dart: just a double
    for d in doubles:
        if d < 100:
            count += 1

    # 2-dart: one non-final + one final double
    for i in range(n):
        for d in doubles:
            if scores[i] + d < 100:
                count += 1

    # 3-dart: two non-final (unordered, i <= j) + one final double
    for i in range(n):
        for j in range(i, n):
            for d in doubles:
                if scores[i] + scores[j] + d < 100:
                    count += 1

    return count

answer = solve()
assert answer == 38182
print(answer)

File: project_euler_unified/problem_109/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    // Build list of all possible dart scores with labels
    // Each dart is a (score, type) pair where type is used for canonical ordering
    // We use index for canonical ordering of non-final darts

    struct Dart {
        int score;
        // For ordering: singles first, then doubles, then trebles
    };

    vector<int> all_darts;   // all 62 dart scores (with distinct identity)
    vector<int> doubles;     // double scores only

    // Singles: S1..S20, S25
    for (int i = 1; i <= 20; i++) all_darts.push_back(i);
    all_darts.push_back(25);

    // Doubles: D1..D20, D25
    for (int i = 1; i <= 20; i++) { all_darts.push_back(2*i); doubles.push_back(2*i); }
    all_darts.push_back(50); doubles.push_back(50);

    // Trebles: T1..T20
    for (int i = 1; i <= 20; i++) all_darts.push_back(3*i);

    int n = all_darts.size(); // 62
    int count = 0;

    // 1-dart checkouts: just a double < 100
    for (int d : doubles) {
        if (d < 100) count++;
    }

    // 2-dart checkouts: one non-final + one double, total < 100
    for (int i = 0; i < n; i++) {
        for (int d : doubles) {
            if (all_darts[i] + d < 100) count++;
        }
    }

    // 3-dart checkouts: two non-final (unordered, i <= j) + one double, total < 100
    for (int i = 0; i < n; i++) {
        for (int j = i; j < n; j++) {
            for (int d : doubles) {
                if (all_darts[i] + all_darts[j] + d < 100) count++;
            }
        }
    }

    cout << count << endl;
    return 0;
}
